packages/compiler/src/transpile.ts

```diff
@@ -1,20 +1,354 @@
+function isChildComponentTrusted(
+  { trustMode, path }: { trustMode: string; path: string },
+  isComponentPathTrusted?: (p: string) => boolean
+) {
+  // child is explicitly trusted by parent or constitutes a new trusted root
+  if (trustMode === TrustMode.Trusted || trustMode === TrustMode.TrustAuthor) {
+    return true;
+  }
+
+  // child is explicitly sandboxed
+  if (trustMode === TrustMode.Sandboxed) {
+    return false;
+  }
+
+  // if the Component is not explicitly trusted or sandboxed, use the parent's
+  // predicate to determine whether the Component should be trusted
+  if (isComponentPathTrusted) {
+    return isComponentPathTrusted(path);
+  }
+
+  return false;
+}
+
+interface TranspileSourceParams {
+  componentPath: string;
+  source: string;
+  isComponentPathTrusted?: (path: string) => boolean;
+}
+
+export function transpileSource({
+  componentPath,
+  source,
+  isComponentPathTrusted,
+}: TranspileSourceParams) {
+  const exports: ModuleExport = { default: '', named: [] };
+  const imports: ModuleImport[] = [];
+  const componentReferences: { [component: string]: ModuleImport } = {};
+  const children: {
+    isTrusted: boolean;
+    path: string;
+    trustMode: string;
+  }[] = [];
+
+  const transformComponents = ({ types: t }: any) => ({
+    visitor: {
+      CallExpression(path: {
+        node: {
+          arguments: [Identifier | StringLiteral, ObjectExpression | undefined];
+          callee: { object: Identifier; property: Identifier };
+        };
+      }) {
+        const {
+          arguments: args,
+          callee: { object, property },
+        } = path.node;
+
+        const isCreateElement =
+          object?.name === '__Preact' && property?.name === 'createElement';
+        const isElement = t.isStringLiteral(args[0]);
+        if (!isCreateElement || isElement) {
+          return;
+        }
+
+        let [Component, props] = args as [Identifier, ObjectExpression];
+
+        if (t.isNullLiteral(props)) {
+          props = t.objectExpression([]);
+          path.node.arguments[1] = props;
+        }
+
+        /**
+         * FIXME referencing `arguments` only works when the Component rendering takes place within
+         *  in a scope in which the correct props is bound to arguments[0], i.e. the Component's root
+         *  scope - in the directly-returned JSX or an arrow function in the Component's root scope.
+         *
+         *  TODO the correct solution is for the parser to inject the __bweMeta reference into the Component
+         *   props argument, but it would also need to be made accessible to the render site
+         *
+         *   TL; DR
+         *   this preserves ancestry:
+         *   import Y from './Y';
+         *   export default function X() { return <Y />; }
+         *
+         *   this doesn't:
+         *   import Y from './Y';
+         *   const renderY = () => <Y />;
+         *   export default function X() { return <>{renderY()}</> }
+         */
```

Comment:
does this merit a dedicated issue? if so, please create one

Reply:
This didn't actually break in the scenarios I expected it to, so it's difficult to articulate the potential impact. I think for now it's a curiosity, but if I come up with a real case I'll write it up.